src/Engine/gramambular2/reading_grid.cpp

```diff
@@ -0,0 +1,601 @@
+  for (size_t i = begin; i <= end; ++i) {
+    spans_[i].removeNodesOfOrLongerThan(loc - i + 1);
+  }
+}
+
+void ReadingGrid::insert(size_t loc, const ReadingGrid::NodePtr& node) {
+  assert(loc < spans_.size());
+  spans_[loc].add(node);
+}
+
+std::string ReadingGrid::combineReading(
+    std::vector<std::string>::const_iterator begin,
+    std::vector<std::string>::const_iterator end) {
+  std::string result;
+  for (auto iter = begin; iter != end;) {
+    result += *iter;
+    ++iter;
+    if (iter != end) {
+      result += separator_;
+    }
+  }
+  return result;
+}
+
+bool ReadingGrid::hasNodeAt(size_t loc, size_t readingLen,
+                            const std::string& reading) {
+  if (loc > spans_.size()) {
+    return false;
+  }
+  const NodePtr& n = spans_[loc].nodeOf(readingLen);
+  if (n == nullptr) {
+    return false;
+  }
+  return reading == n->reading();
+}
+
+void ReadingGrid::update() {
+  size_t begin =
+      (cursor_ <= kMaximumSpanLength) ? 0 : cursor_ - kMaximumSpanLength;
+  size_t end = cursor_ + kMaximumSpanLength;
+  if (end > readings_.size()) {
+    end = readings_.size();
+  }
+
+  for (size_t pos = begin; pos < end; pos++) {
+    for (size_t len = 1; len <= kMaximumSpanLength && pos + len <= end; len++) {
+      std::string combinedReading =
+          combineReading(readings_.begin() + static_cast<ptrdiff_t>(pos),
+                         readings_.begin() + static_cast<ptrdiff_t>(pos + len));
+
+      if (!hasNodeAt(pos, len, combinedReading)) {
+        auto unigrams = lm_.getUnigrams(combinedReading);
+        if (unigrams.empty()) {
+          continue;
+        }
+
+        insert(pos, std::make_shared<Node>(combinedReading, len, unigrams));
+      }
+    }
+  }
+}
+
+std::vector<ReadingGrid::NodeInSpan> ReadingGrid::overlappingNodesAt(
+    size_t loc) {
+  std::vector<ReadingGrid::NodeInSpan> results;
+
+  if (spans_.empty() || loc >= spans_.size()) {
+    return results;
+  }
+
+  // First, get all nodes from the span at location.
+  for (size_t i = 1, len = spans_[loc].maxLength(); i <= len; ++i) {
+    NodePtr ptr = spans_[loc].nodeOf(i);
+    if (ptr != nullptr) {
+      ReadingGrid::NodeInSpan element{.node = std::move(ptr), .spanIndex = loc};
+      results.emplace_back(std::move(element));
+    }
+  }
+
+  size_t begin = loc - std::min(loc, kMaximumSpanLength - 1);
+  for (size_t i = begin; i < loc; ++i) {
+    size_t beginLen = loc - i + 1;
+    size_t endLen = spans_[i].maxLength();
+    for (size_t j = beginLen; j <= endLen; ++j) {
+      NodePtr ptr = spans_[i].nodeOf(j);
+      if (ptr != nullptr) {
+        ReadingGrid::NodeInSpan element{.node = std::move(ptr), .spanIndex = i};
+        results.emplace_back(std::move(element));
+      }
+    }
+  }
+
+  return results;
+}
+
+std::string ReadingGrid::Node::value() const {
+  return unigrams_.empty() ? "" : unigramIter_->value();
+}
+
+double ReadingGrid::Node::score() const {
+  if (unigrams_.empty()) {
+    return 0;
+  }
+
```

Comment:
> I remember that we won't override the symbols (nodes with scores <= -8). If a user type "好" and then select "👍", "👍" will appear next time when he or she types "好", and it is annoying. It looks like the logic is not implemented in gramambular 2.

Yes, but that is because of UserOverrideModel, and you implemented a rule in KeyHandler to prevent UOM from providing a suggested override in such cases. I've carried over that rule in #58.